sympy/sets/fancysets.py

```diff
@@ -404,52 +471,178 @@ def _intersect(self, other):
         return None
 
     def _contains(self, other):
-        if (((self.start - other)/self.step).is_integer or
-            ((self.stop - other)/self.step).is_integer):
-            return _sympify(other >= self.inf and other <= self.sup)
-        elif (((self.start - other)/self.step).is_integer is False and
-            ((self.stop - other)/self.step).is_integer is False):
+        if other.is_integer is None:
+            return
+        if other.is_integer is False:
+            return S.false
+        ref = self.start if self.start.is_finite else self.stop
+        if (ref - other) % self.step:  # off sequence
             return S.false
+        return _sympify(other >= self.inf and other <= self.sup)
 
     def __iter__(self):
-        if self.start is S.NegativeInfinity:
-            i = self.stop - self.step
-            step = -self.step
-        else:
-            i = self.start
-            step = self.step
+        if self.start in [S.NegativeInfinity, S.Infinity]:
+            raise ValueError("Cannot iterate over Range with infinite start")
+
+        i = self.start
+        step = self.step
 
-        while(i < self.stop and i >= self.start):
+        while True:
+            if (step > 0 and not (self.start <= i < self.stop)) or \
+               (step < 0 and not (self.stop < i <= self.start)):
+                break
             yield i
             i += step
 
+
     def __len__(self):
-        return (self.stop - self.start)//self.step
+        dif = self.stop - self.start
+        if dif.is_infinite:
+            raise ValueError(
+                "Use .size to get the length of an infinite Range")
+        return abs(dif//self.step)
+
+    @property
+    def size(self):
+        dif = self.stop - self.start
+        if dif.is_infinite:
+            return S.Infinity
+        return abs(dif//self.step)
 
     def __nonzero__(self):
         return True
 
     __bool__ = __nonzero__
 
-    def _ith_element(self, i):
-        return self.start + i*self.step
-
-    @property
-    def _last_element(self):
-        if self.stop is S.Infinity:
-            return S.Infinity
-        elif self.start is S.NegativeInfinity:
-            return self.stop - self.step
+    def __getitem__(self, i):
+        from sympy.functions.elementary.complexes import sign
+        from sympy.functions.elementary.integers import ceiling, floor
+        ooslice = "cannot slice from infinite end"
+        zerostep = "slice step cannot be zero"
+        # if we had to take every other element in the following
+        # oo, ..., 6, 4, 2, 0
+        # we might get oo, ..., 4, 0 or oo, ..., 6, 2
+        ambiguous = "cannot unambiguously re-stride from infinite end"
+        oo_element = "can't represent [%s] with Range"
+        if isinstance(i, slice):
+            if self.size.is_finite:
+                start, stop, step = i.indices(self.size)
+                n = ceiling((stop - start)/step)
```

Comment:
Oh, I guess you mean by `//` that perhaps we should be using integer division. But then ceiling would not be necessary. In this file, `division` has been imported so we are truly moving `n` up. It could be done using divmod, too. What is here works, however, and is transparent in terms of what is being done.